src/server/services/vault.service.ts

```diff
@@ -0,0 +1,567 @@
+      status: VaultItemStatus.Pending, // Resets to pending.
+    },
+  });
+};
+
+export const getPaginatedVaultItems = async (
+  input: GetPaginatedVaultItemsSchema & { userId?: number }
+) => {
+  const { limit = DEFAULT_PAGE_SIZE, page } = input || {};
+  const { take, skip } = getPagination(limit, page);
+
+  const where: Prisma.VaultItemFindManyArgs['where'] = {
+    vaultId: input.userId,
+  };
+  const orderBy: Prisma.VaultItemFindManyArgs['orderBy'] = {};
+
+  if (input.sort === VaultSort.RecentlyAdded) {
+    orderBy.addedAt = 'desc';
+  } else if (input.sort === VaultSort.RecentlyCreated) {
+    orderBy.createdAt = 'desc';
+  } else if (input.sort === VaultSort.ModelName) {
+    orderBy.modelName = 'asc';
+  } else if (input.sort === VaultSort.ModelSize) {
+    orderBy.modelSizeKb = 'desc';
+  }
+
+  if (input.query) {
+    where.OR = [
+      {
+        modelName: {
+          contains: input.query,
+          mode: 'insensitive',
+        },
+      },
+      {
+        versionName: {
+          contains: input.query,
+          mode: 'insensitive',
+        },
+      },
+      {
+        creatorName: {
+          contains: input.query,
+          mode: 'insensitive',
+        },
+      },
+    ];
+  }
+
+  if (input.types && input.types.length) {
+    where.type = { in: input.types };
+  }
+
+  if (input.categories && input.categories.length) {
+    where.category = { in: input.categories };
+  }
+
+  if (input.baseModels && input.baseModels.length) {
+    where.baseModel = { in: input.baseModels };
+  }
+
+  if (input.dateCreatedFrom) {
+    where.createdAt = { gte: input.dateCreatedFrom };
+  }
+
+  if (input.dateCreatedTo) {
+    where.createdAt = { lte: input.dateCreatedTo };
+  }
+
+  if (input.dateAddedFrom) {
+    where.addedAt = { gte: input.dateAddedFrom };
+  }
+
+  if (input.dateAddedTo) {
+    where.addedAt = { lte: input.dateAddedTo };
+  }
+
+  const items = await dbRead.vaultItem.findMany({
+    where,
+    take,
+    skip,
+    orderBy,
+  });
+
+  const itemsWithCoverImages = await Promise.all(
+    items.map(async (item) => {
+      const { url } =
+        item.status === VaultItemStatus.Stored
+          ? await getGetUrlByKey(
+              constants.vault.keys.cover
+                .replace(':modelVersionId', item.modelVersionId.toString())
+                .replace(':userId', item.vaultId.toString()),
+              { bucket: env.S3_VAULT_BUCKET }
+            )
+          : { url: null };
+
+      return {
+        ...item,
+        coverImageUrl: url,
+        files: (item.files ?? []) as VaultItemFilesSchema,
+      };
+    })
+  );
+
+  const count = await dbRead.vaultItem.count({ where });
+
+  return getPagingData({ items: itemsWithCoverImages, count: (count as number) ?? 0 }, limit, page);
+};
+
+export const isModelVersionInVault = async ({
+  userId,
+  modelVersionId,
+}: VaultItemsAddModelVersionSchema & {
+  userId: number;
+}) => {
+  const exists = await dbRead.vaultItem.findFirst({
```

Comment:
Would count be more lightweight?
```suggestion
  const exists = await dbRead.vaultItem.count({
```

Reply:
No, count requires grouping - should be less performant if anything